Tools/SwiftBrowser/Source/BrowserViewModel.swift

```diff
@@ -39,17 +39,87 @@ extension PDF: Transferable {
     }
 }
 
+@MainActor
+final class DialogPresenter: DialogPresenting {
+    struct Dialog: Hashable, Identifiable, Sendable {
+        enum Configuration: Sendable {
+            case alert(String, @Sendable () -> Void)
+            case confirm(String, @Sendable (WebPage_v0.JavaScriptConfirmResult) -> Void)
+            case prompt(String, defaultText: String?, @Sendable (WebPage_v0.JavaScriptPromptResult) -> Void)
+        }
+
+        let id = UUID()
+        let configuration: Configuration
+
+        func hash(into hasher: inout Hasher) {
+            self.id.hash(into: &hasher)
+        }
+
+        static func == (lhs: Dialog, rhs: Dialog) -> Bool {
+            lhs.id == rhs.id
+        }
+    }
+
+    struct FilePicker {
+        let allowsMultipleSelection: Bool
+        let allowsDirectories: Bool
+        let completion: (WebPage_v0.FileInputPromptResult) -> Void
+    }
+
+    weak var owner: BrowserViewModel? = nil
+
+    func handleJavaScriptAlert(message: String, initiatedBy frame: WebPage_v0.FrameInfo) async {
+        await withCheckedContinuation { continuation in
+            owner?.currentDialog = Dialog(configuration: .alert(message) {
```

Comment:
Super nit, but you should be able to just pass this as an unapplied function rather than making a closure just to call it like a thunk:
```suggestion
            owner?.currentDialog = Dialog(configuration: .alert(message, continuation.resume))
```